### DISCUSSION TITLE
指定下载 AI 字幕的方法

### DISCUSSION THREAD

**Discussion by WakelessSloth56** [the1812/Bilibili-Evolved#4795] [分享 / Share]
最近想要下载 AI 字幕时想到：https://github.com/the1812/Bilibili-Evolved/issues/4319#issuecomment-1644844670 (@the1812) 

> b 站选择 AI 字幕不会更新到 localStorage 中存储的字幕语言，脚本匹配到的仍然是非 AI 字幕

经过研究，可通过播放器 API 来获取字幕设置：

```js
player.__core().subtitleStore.state.lan
```

例如：选择 `中文（自动生成）` 时，此值为 `ai-zh`。

字幕地址，字幕显示设置等也可从此处获得，详情我还是整理在了 #4341 下。

---

`player.__core()` 下有很多功能的方法，以前就有发现，但是因为本身 `player` 下方法就挺丰富的就没怎么重视，最近想要下载 AI 字幕时想起来 Bilibili-Evolved 这个存在了好久的问题，才有想起这个东西来。以后有时间再整理其他的。

**Comment by the1812** [comment]
`player.__core().subtitleStore` 在部分播放器 (例如稍后再看) 中是 `undefined`

**Reply by WakelessSloth56** [reply]
啊这……我写这个的时候还在的……看了一下版本信息，这是前几天刚更新掉的啊……😕
```
/*! @metadata nano: 4.8.34-3d75f24d (2024-07-11T14:11:05+08:00) */
```
其他播放页的还没改：
```
/*! @metadata nano: 4.8.33-460db19e (2024-07-02T15:19:58+08:00) */
```

**Comment by CaoMeiYouRen** [comment]
有关这个问题，是否有进一步的研究？
虽然找到了接口，但不知道参数从哪来的。
```
https://aisubtitle.hdslb.com/bfs/ai_subtitle/prod/1131474573087902590697231248f48430f23f3abe7b6fc6f162673b77?auth_key=xx
https://aisubtitle.hdslb.com/bfs/subtitle/54d6758e76e8eb6affae9f6549ee736c241b2432.json?auth_key=xx

```

**Reply by WakelessSloth56** [reply]
请仔细看一下这个讨论和他引用的议题，问题并不是拿不到 AI 字幕，而且不能确定当前播放器是不是选择了 AI 字幕。

接口 `api.bilibili.com/x/player/wbi/v2` 是可以拿到包括 AI 字幕在内的所有字幕的地址的，只是例如视频有两个字幕，如果想要下载 AI 字幕，但是因为

> b 站选择 AI 字幕不会更新到 localStorage 中存储的字幕语言，脚本匹配到的仍然是非 AI 字幕

所以只能下载到非 AI 字幕的那一个。

而我当时在此讨论提出的方法，是通过播放器接口获取当前选择的字幕是哪一个，但是就在我写这个讨论后不久，B 站的播放器就升级了，不再暴露这些方法了。

**Reply by CaoMeiYouRen** [reply]
> 请仔细看一下这个讨论和他引用的议题，问题并不是拿不到 AI 字幕，而且不能确定当前播放器是不是选择了 AI 字幕。
> 
> 接口 `api.bilibili.com/x/player/wbi/v2` 是可以拿到包括 AI 字幕在内的所有字幕的地址的，只是例如视频有两个字幕，如果想要下载 AI 字幕，但是因为
> 
> > b 站选择 AI 字幕不会更新到 localStorage 中存储的字幕语言，脚本匹配到的仍然是非 AI 字幕
> 
> 所以只能下载到非 AI 字幕的那一个。
> 
> 而我当时在此讨论提出的方法，是通过播放器接口获取当前选择的字幕是哪一个，但是就在我写这个讨论后不久，B 站的播放器就升级了，不再暴露这些方法了。

感谢回复。

如果是这样的话，是否可以考虑让用户手动选择需要下载的字幕；或者更加暴力一点，下载所有字幕。

然后就是，现在似乎非 AI 字幕也下载不了了

**Reply by WakelessSloth56** [reply]
我已经提出新的讨论了（#4927）。

无法下载字幕的情况我还没有遇到。

**Reply by CaoMeiYouRen** [reply]
> 我已经提出新的讨论了（#4927）。
> 
> 无法下载字幕的情况我还没有遇到。

破案了，是下载到了旧版本的脚本，新版本没有问题。
